File: Funções_Java/G250_easyPrinter/src/main/java/com/gertec/G250/functions/printCouponNFCe.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.gertec.G250.functions;

import gnu.io.NRSerialPort;
import java.io.File;
import java.io.IOException;
import java.net.Socket;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.swing.text.MaskFormatter;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

/**
 *
 * @author adels
 */
public class printCouponNFCe {
    
    private static final byte[] RESTART = {0x1b, 0x40}; // limpa os dados e reset as configurações da impressora

    
    public static void printCouponNFCe(String path, int type, Socket socket) throws IOException, Exception {
        sendBinBuffer.sendBinBuffer(RESTART, socket);
        printText.printText(printCouponNFCeFactory(path, type), socket);
        System.out.println(printCouponNFCeFactory(path, type));
    }

    public static void printCouponNFCe(String path, int type, NRSerialPort serial) throws IOException, Exception {
        sendBinBuffer.sendBinBuffer(RESTART, serial);
        printText.printText(printCouponNFCeFactory(path, type), serial);
        //System.out.println(printCouponNFCeFactory(path, type));
    }
    
    public static String printCouponNFCeFactory(String path, int type) {
        StringBuilder buffer = new StringBuilder();
        String nfceType, nfceObs;
        switch (type) {
            case 1:
                nfceType = "Documento Auxiliar da Nota Fiscal de Consumidor Eletrônica";
                nfceObs = "Não permite aproveitamento de crédito do ICMS";
                break;
            case 2:
                nfceType = "EMITIDA EM CONTIGÊNCIA";
                nfceObs = "Pendente de autorização";
                break;
            case 3:
                nfceType = "CANCELAMENTO";
                nfceObs = "Pendente de autorização";
                break;
            default:
                throw new IllegalArgumentException("Tipo de Coupon inválido. Os valores válidos são 1 (NFCE)"
                    + ", 2 (Contigência) ou 3 (Cancelamento)");
        }
        try{
            File file = new File (path);
            DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
            DocumentBuilder db = dbf.newDocumentBuilder();
            Document document = db.parse(file);
            document.getDocumentElement().normalize();
            NodeList nList = document.getElementsByTagName("nfeProc");
            for(int temp = 0; temp < nList.getLength(); temp++){
                Node nNode = nList.item(temp);
                Element eElement = (Element) nNode;
                if (nNode.getNodeType() == Node.ELEMENT_NODE){
                    
                    //cabeçalho
                    String cnpj = eElement.getElementsByTagName("CNPJ").item(0).getTextContent();
                    MaskFormatter mask = new MaskFormatter("###.###.###/####-##");
                    mask.setValueContainsLiteralCharacters(false);
                    buffer.append("<ce><c>").append("CNPJ:").append(mask.valueToString(cnpj)).append("</lf>");
                    buffer.append("<ne>").append(eElement.getElementsByTagName("xNome").item(0).getTextContent()).append("</ne></lf>");
                    buffer.append(eElement.getElementsByTagName("xLgr").item(0).getTextContent());
                    buffer.append(", ").append(eElement.getElementsByTagName("nro").item(0).getTextContent());
                    buffer.append(", ").append(eElement.getElementsByTagName("xBairro").item(0).getTextContent());
                    buffer.append(", ").append(eElement.getElementsByTagName("xMun").item(0).getTextContent());
                    buffer.append(", ").append(eElement.getElementsByTagName("UF").item(0).getTextContent()).append("</lf></lf>");
                    buffer.append("<ne>").append(nfceType).append("</ne></lf>");
                    buffer.append(nfceObs).append("</lf></lf>");
                    
                    //Informações dos produtos
                    buffer.append("DETALHE DA VENDA").append("</ce><c></lf>");
                    buffer.append("________________________________________________________________</lf>");
                    buffer.append("<ne>Código  Descrição                Qtde UN      Vl Unit   Vl Total</ne>").append("</ne></lf>");
                    NodeList produtos = eElement.getElementsByTagName("prod");
                    for(int i = 0; i < produtos.getLength(); i++){
                        String cProd = produtos.item(i).getChildNodes().item(0).getTextContent();
                        buffer.append(cProd.substring(cProd.length()-6, cProd.length())).append(String.format("%2s",""));
                        String prodName = produtos.item(i).getChildNodes().item(2).getTextContent();
                        if (prodName.length() > 20){
                            buffer.append(prodName.substring(0, 20)).append(String.format("%4s",""));
                        }else
                            buffer.append(prodName).append(String.format("%"+(24-prodName.length())+"s",""));
                        String qCom = produtos.item(i).getChildNodes().item(7).getTextContent();
                        buffer.append(qCom).append(String.format("%"+(8-qCom.length())+"s",""));
                        String uCom = produtos.item(i).getChildNodes().item(6).getTextContent();
                        buffer.append(uCom).append(String.format("%"+(6-uCom.length())+"s",""));
                        String vUnCom = produtos.item(i).getChildNodes().item(8).getTextContent();
                        buffer.append(vUnCom).append(String.format("%"+(13-vUnCom.length())+"s",""));
                        String vProd = produtos.item(i).getChildNodes().item(9).getTextContent();
                        buffer.append(String.format("%5s",vProd)).append("</lf>");
                        
                        int remain = prodName.length() - 20;
                        while (remain > 0) {                            
                            if (remain > 20) {
                                buffer.append(String.format("%8s","")).append(prodName.substring(prodName.length() - remain, prodName.length() - remain + 20)).append("</lf>");
                            }else
                                buffer.append(String.format("%8s","")).append(prodName.substring(prodName.length() - remain, prodName.length())).append("</lf>");
                            remain -= 20;
                        }
                    }
                    buffer.append("________________________________________________________________</lf>");
                    String qtdItens = eElement.getElementsByTagName("indTot").item(0).getTextContent();
                    buffer.append("Qtde. total de Itens:").append(String.format("%"+(43)+"s",qtdItens)).append("</lf>");
                    String vTotal = eElement.getElementsByTagName("vPag").item(0).getTextContent();
                    buffer.append("Valor Total R$:").append(String.format("%"+(64-15)+"s",vTotal)).append("</lf>");
                    String desconto = eElement.getElementsByTagName("vDesc").item(0).getTextContent();
                    buffer.append("Desconto R$:").append(String.format("%"+(64-11)+"s",desconto)).append("</lf>");
                    String frete = eElement.getElementsByTagName("vFrete").item(0).getTextContent();
                    buffer.append("Frete R$:").append(String.format("%"+(64-9)+"s",frete)).append("</lf>");
                    String vPagar = eElement.getElementsByTagName("vNF").item(0).getTextContent();
                    buffer.append("<ne>").append("Valor a Pagar R$:").append(String.format("%"+(64-17)+"s",vPagar)).append("</lf></ne>");
                    buffer.append("________________________________________________________________</lf>");
                    
                    //Informações de Totais do DANFE NFC-e
                    String tPag = eElement.getElementsByTagName("tPag").item(0).getTextContent();
                    String typeOfPayment;
                    switch (tPag) {
                        case "01": typeOfPayment = "Dinheiro"; break;
                        case "02": typeOfPayment = "Cheque"; break;
                        case "03": typeOfPayment = "Cartão de Crédito"; break;
                        case "04": typeOfPayment = "Cartão de Débito"; break;
                        case "05": typeOfPayment = "Crédito Loja"; break;
                        case "10": typeOfPayment = "Vale Alimentação"; break;
                        case "11": typeOfPayment = "Vale Refeição"; break;
                        case "12": typeOfPayment = "Vale Presente"; break;
                        case "13": typeOfPayment = "Vale Combustível"; break;
                        case "14": typeOfPayment = "Duplicata Mercantil"; break;
                        case "15": typeOfPayment = "Boleto Bancário"; break;
                        case "90": typeOfPayment = "Sem Pagamento"; break;
                        case "99": typeOfPayment = "Outros"; break;
                        default:
                            return "Invalid option for TYPE OF PAYMENT.";
                    }
                    
                    buffer.append("FORMA DE PAGAMENTO").append(String.format("%"+(64-16)+"s", "VALOR PAGO")).append("</lf>");
                    //Verificar esse vPag
                    String vPag = eElement.getElementsByTagName("vPag").item(0).getTextContent();
                    buffer.append(typeOfPayment).append(String.format("%"+(64-typeOfPayment.length())+"s", vPag)).append("</lf>");
                    buffer.append("________________________________________________________________</lf></lf>");
                    buffer.append("<ce><ne>Consulte pela Chave de Acesso em").append("</lf></ne>");
                    buffer.append(eElement.getElementsByTagName("urlChave").item(0).getTextContent()).append("</lf>");
                    mask.setMask("#### #### #### #### #### #### #### #### #### #### ####");
                    buffer.append(mask.valueToString(eElement.getElementsByTagName("chNFe").item(0).getTextContent())).append("<c></lf>");
                    buffer.append("<qr>5,48,").append(eElement.getElementsByTagName("qrCode").item(0).getTextContent()).append("</qr></lf>");
                    buffer.append("Protocolo de Autorização</lf>");
                    buffer.append(eElement.getElementsByTagName("nProt").item(0).getTextContent()).append("</lf>");
                    buffer.append("Data de Emissão: ").append(eElement.getElementsByTagName("dhEmi").item(0).getTextContent()).append("</ce></lf></gui>");
                }
            }
            
            
           
        } catch (ParserConfigurationException ex) {
            Logger.getLogger(printCouponNFCe.class.getName()).log(Level.SEVERE, null, ex);
        } catch (SAXException ex) {
            Logger.getLogger(printCouponNFCe.class.getName()).log(Level.SEVERE, null, ex);
        } catch (IOException ex) {
            Logger.getLogger(printCouponNFCe.class.getName()).log(Level.SEVERE, null, ex);
        } catch (Exception ex) {
            Logger.getLogger(printCouponNFCe.class.getName()).log(Level.SEVERE, null, ex);
        }
        
        return buffer.toString();
    }
    
}

File: C#/printCouponNFCe.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml;

namespace GertecEasyPrint
{
    internal class printCouponNFCe {
        public static String PrintCouponNFCe(String filePath, int type) {
            if (!File.Exists(filePath)) {
                return "File not found.";
            }

            String nfceType, nfceObs;
            String spaceChar = "";

            switch (type) {
                case 1:
                    nfceType = "Documento Auxiliar da Nota Fiscal de Consumidor Eletrônica";
                    nfceObs = "Não permite aproveitamento de crédito do ICMS";
                    break;
                case 2:
                    nfceType = "EMITIDA EM CONTIGÊNCIA";
                    nfceObs = "Pendente de autorização";
                    break;
                case 3:
                    nfceType = "DOCUMENTO CANCELADO";
                    nfceObs =  "-------------------";
                    break;
                default:
                    return "Invalid option for NFCE TYPE.";
            }


            XmlDocument xml = new XmlDocument();
            xml.Load(filePath);

            //Divisão I - Informações do Cabeçalho
            String nfceBuilder = "";
            nfceBuilder += printBITMAP.EncodeBitmap(@"C:\logogertec.jpg");
            nfceBuilder += "\x0a\x0a"; //pular linha
            nfceBuilder += "\x1b\x4d\x01"; //fonte pequena
            nfceBuilder += "\x1b\x32"; //espacamento entre linhas
            nfceBuilder += "\x1b\x61\x01"; //centralizar
            String formatCNPJ = xml.GetElementsByTagName("CNPJ")[0].InnerText;
            nfceBuilder += "CNPJ:" + Convert.ToUInt64(formatCNPJ).ToString(@"00\.000\.000\/0000\-00") + "\x0a";
            nfceBuilder += "\x1b\x45\x01"; //negrito on
            nfceBuilder += xml.GetElementsByTagName("xNome")[0].InnerText + "\x0a";
            nfceBuilder += "\x1b\x45\x00"; //negrito off
            nfceBuilder += xml.GetElementsByTagName("xLgr")[0].InnerText + ", ";
            nfceBuilder += xml.GetElementsByTagName("nro")[0].InnerText + ", ";
            nfceBuilder += xml.GetElementsByTagName("xBairro")[0].InnerText + ", ";
            nfceBuilder += xml.GetElementsByTagName("xMun")[0].InnerText + ", ";
            nfceBuilder += xml.GetElementsByTagName("UF")[0].InnerText + "\x0a";
            nfceBuilder += "\x1b\x45\x01"; //negrito
            nfceBuilder += "\x0a" + nfceType;
            nfceBuilder += "\x1b\x45\x00"; //negrito off
            nfceBuilder += "\x0a" + nfceObs;

            //Divisão II – Informações de detalhes de produtos/serviços
            nfceBuilder += "\x1b\x45\x01";
            nfceBuilder += "\x0a\x0a" + "DETALHE DA VENDA" + "\x0a";
            nfceBuilder += "\x1b\x61\x00";
            nfceBuilder += "________________________________________________________________";
            nfceBuilder += "Código  Descrição              Qtde UN      Vl Unit     Vl Total" + "\x0a";
            nfceBuilder += "\x1b\x45\x00";
            int countItems = 0;
            XmlNodeList productList = xml.GetElementsByTagName("prod");
            for (int i = 0; i < productList.Count; i++) {
                String cProd = xml.GetElementsByTagName("cProd")[i].InnerText;
                String xProd = xml.GetElementsByTagName("xProd")[i].InnerText;
                String qCom = xml.GetElementsByTagName("qCom")[i].InnerText;
                String uCom = xml.GetElementsByTagName("uCom")[i].InnerText;
                String vUnCom = xml.GetElementsByTagName("vUnCom")[i].InnerText;
                String vProd = xml.GetElementsByTagName("vProd")[i].InnerText;

                String buildFirstLine = "";
                if (cProd.Length <= 6) {
                    buildFirstLine += cProd + spaceChar.PadLeft(8 - cProd.Length);
                } else {
                    buildFirstLine += cProd.Substring(cProd.Length - 6, 6) + "  ";
                }

                bool stop = false;
                bool first = false;
                do {
                    if (xProd.Length <= 20) {
                        buildFirstLine += xProd + spaceChar.PadLeft(23 - xProd.Length);
                        stop = true;
                        if (first == false) {
                            buildFirstLine += qCom + " ";
                            buildFirstLine += uCom + spaceChar.PadLeft(10 - qCom.Length);
                            buildFirstLine += vUnCom + spaceChar.PadLeft(20 - (vUnCom.Length + vProd.Length)); ;
                            buildFirstLine += vProd;
                        }
                    } else {
                        if (first == false) {
                            buildFirstLine += xProd.Substring(0, 20) + spaceChar.PadLeft(3);
                            xProd = xProd.Substring(20);
                            buildFirstLine += qCom + " ";
                            buildFirstLine += uCom + spaceChar.PadLeft(10 - qCom.Length);
                            buildFirstLine += vUnCom + spaceChar.PadLeft(20 - (vUnCom.Length + vProd.Length)); ;
                            buildFirstLine += vProd + spaceChar.PadLeft(8);
                            first = true;
                        } else {
                            buildFirstLine += xProd.Substring(0, 20) + spaceChar.PadLeft(44);
                            xProd = xProd.Substring(20);
                        }
                    }
                } while (stop != true);

                nfceBuilder += buildFirstLine;
                nfceBuilder += "\x0a";
                countItems++;
            }

            //Divisão III – Informações de Totais do DANFE NFC-e
            nfceBuilder += "________________________________________________________________" + "\x0a";
            nfceBuilder += "Qtde. total de itens: " + spaceChar.PadLeft(42 - countItems.ToString().Length) + countItems + "\x0a";
            String amount = xml.GetElementsByTagName("vProd")[countItems].InnerText;
            nfceBuilder += "Valor Total R$: " + spaceChar.PadLeft(48 - amount.Length) + amount + "\x0a";
            String discount = xml.GetElementsByTagName("vDesc")[0].InnerText;
            nfceBuilder += "Desconto R$: " + spaceChar.PadLeft(51 - discount.Length) + discount + "\x0a";
            nfceBuilder += "\x1b\x45\x01";
            String amountToPay = xml.GetElementsByTagName("vNF")[0].InnerText;
            nfceBuilder += "Valor a Pagar R$: " + spaceChar.PadLeft(46 - amountToPay.Length) + amountToPay + "\x0a";
            nfceBuilder += "\x1b\x45\x00";
            nfceBuilder += "________________________________________________________________" + "\x0a";
            nfceBuilder += "FORMA DE PAGAMENTO                                    VALOR PAGO" + "\x0a";
            String typeOfPayment = xml.GetElementsByTagName("tPag")[0].InnerText;

            switch (typeOfPayment) {
                case "01": typeOfPayment = "Dinheiro"; break;
                case "02": typeOfPayment = "Cheque"; break;
                case "03": typeOfPayment = "Cartão de Crédito"; break;
                case "04": typeOfPayment = "Cartão de Débito"; break;
                case "05": typeOfPayment = "Crédito Loja"; break;
                case "10": typeOfPayment = "Vale Alimentação"; break;
                case "11": typeOfPayment = "Vale Refeição"; break;
                case "12": typeOfPayment = "Vale Presente"; break;
                case "13": typeOfPayment = "Vale Combustível"; break;
                case "14": typeOfPayment = "Duplicata Mercantil"; break;
                case "15": typeOfPayment = "Boleto Bancário"; break;
                case "90": typeOfPayment = "Sem Pagamento"; break;
                case "99": typeOfPayment = "Outros"; break;
                default:
                    return "Invalid option for TYPE OF PAYMENT.";
            }

            String amountPaid = xml.GetElementsByTagName("vPag")[0].InnerText;
            nfceBuilder += typeOfPayment + spaceChar.PadLeft(64 - amountPaid.Length - typeOfPayment.Length) + amountPaid;
            nfceBuilder += "________________________________________________________________" + "\x0a\x0a";
            //Divisão IV – Informações da consulta via chave de acesso
            nfceBuilder += "\x1b\x61\x01";
            nfceBuilder += "\x1b\x45\x01";
            nfceBuilder += "Consulte pela Chave de Acesso em" + "\x0a";
            nfceBuilder += "\x1b\x45\x00";
            nfceBuilder += xml.GetElementsByTagName("urlChave")[0].InnerText + "\x0a";
            String nfeKey = xml.GetElementsByTagName("chNFe")[0].InnerText;
                      
            for(int i = 0; i < 11; i++) {
                nfceBuilder += nfeKey.Substring(i*4, 4) + " ";
            }
            nfceBuilder += "\x0a\x0a";
            //Divisão V – Informações da consulta via QR Code
            nfceBuilder += "Protocolo de Autorização: " + xml.GetElementsByTagName("nProt")[0].InnerText + "\x0a";
            String AutDate = xml.GetElementsByTagName("dhRecbto")[0].InnerText;
            nfceBuilder += "Data de Autorização: " + Convert.ToDateTime(AutDate) + "\x0a";

            if(type == 2) {
                nfceBuilder += "\x1b\x45\x01"; //negrito
                nfceBuilder += "\x0a" + nfceType;
                nfceBuilder += "\x1b\x45\x00"; //negrito off
                nfceBuilder += "\x0a" + nfceObs + "\x0a";
            }

            String qrCode = xml.GetElementsByTagName("qrCode")[0].InnerText;
            nfceBuilder += printQRCODE.EncodeQRCODE(qrCode);
            nfceBuilder += "\x0a\x0a";
            nfceBuilder += "\x0a\x0a";
            return nfceBuilder;
        }
    }
}
